File: sdk/elixir/test/rocksky/filter_test.exs
defmodule Rocksky.FilterTest do
  use ExUnit.Case, async: true

  alias Rocksky.Filter

  doctest Rocksky.Filter

  describe "comparisons" do
    test "eq with a safe string stays bare" do
      assert Filter.eq(:artist, "Radiohead") |> Filter.build() == "artist==Radiohead"
    end

    test "eq quotes a string containing a space" do
      assert Filter.eq(:artist, "Daft Punk") |> Filter.build() == "artist==\"Daft Punk\""
    end

    test "eq escapes embedded double quotes" do
      assert Filter.eq(:title, ~s(He said "hi")) |> Filter.build() ==
               ~s(title=="He said \\"hi\\"")
    end

    test "eq keeps * wildcards unquoted" do
      assert Filter.eq(:artist, "Daft*") |> Filter.build() == "artist==Daft*"
    end

    test "ne" do
      assert Filter.ne(:artist, "Eminem") |> Filter.build() == "artist!=Eminem"
    end

    test "gt renders integers without decimals" do
      assert Filter.gt(:duration, 200_000) |> Filter.build() == "duration=gt=200000"
    end

    test "ge" do
      assert Filter.ge(:year, 2000) |> Filter.build() == "year=ge=2000"
    end

    test "lt" do
      assert Filter.lt(:trackNumber, 5) |> Filter.build() == "trackNumber=lt=5"
    end

    test "le" do
      assert Filter.le(:year, 1999) |> Filter.build() == "year=le=1999"
    end

    test "eq with a boolean" do
      assert Filter.eq(:liked, true) |> Filter.build() == "liked==true"
    end

    test "dotted atom selectors" do
      assert Filter.eq(:"track.artist", "Daft Punk") |> Filter.build() ==
               "track.artist==\"Daft Punk\""
    end

    test "binary fields are accepted too" do
      assert Filter.eq("artist", "Radiohead") |> Filter.build() == "artist==Radiohead"
      assert Filter.gt("track.duration", 200_000) |> Filter.build() == "track.duration=gt=200000"
    end
  end

  describe "in / out" do
    test "is_in" do
      assert Filter.is_in(:genre, ["house", "electro"]) |> Filter.build() ==
               "genre=in=(house,electro)"
    end

    test "is_out quotes values with reserved characters" do
      assert Filter.is_out(:genre, ["hip hop"]) |> Filter.build() == ~s|genre=out=("hip hop")|
    end

    test "is_in with an empty list raises" do
      assert_raise ArgumentError, fn -> Filter.is_in(:genre, []) end
    end

    test "is_out with an empty list raises" do
      assert_raise ArgumentError, fn -> Filter.is_out(:genre, []) end
    end
  end

  describe "null checks" do
    test "is_null" do
      assert Filter.is_null(:uri) |> Filter.build() == "uri==null"
    end

    test "is_not_null" do
      assert Filter.is_not_null(:uri) |> Filter.build() == "uri!=null"
    end
  end

  describe "combinators" do
    test "and_ joins with ;" do
      assert Filter.eq(:artist, "Radiohead")
             |> Filter.and_(Filter.gt(:duration, 200_000))
             |> Filter.build() == "artist==Radiohead;duration=gt=200000"
    end

    test "or_ joins with ," do
      assert Filter.eq(:artist, "Radiohead")
             |> Filter.or_(Filter.eq(:artist, "Muse"))
             |> Filter.build() == "artist==Radiohead,artist==Muse"
    end

    test "and_ parenthesizes an or left operand" do
      assert Filter.eq(:artist, "Radiohead")
             |> Filter.or_(Filter.eq(:artist, "Muse"))
             |> Filter.and_(Filter.gt(:duration, 200_000))
             |> Filter.build() == "(artist==Radiohead,artist==Muse);duration=gt=200000"
    end

    test "and_ parenthesizes an or right operand" do
      assert Filter.eq(:artist, "Radiohead")
             |> Filter.and_(Filter.eq(:genre, "house") |> Filter.or_(Filter.eq(:genre, "electro")))
             |> Filter.build() == "artist==Radiohead;(genre==house,genre==electro)"
    end

    test "or_ never parenthesizes" do
      assert Filter.eq(:artist, "Radiohead")
             |> Filter.and_(Filter.gt(:duration, 200_000))
             |> Filter.or_(Filter.eq(:genre, "house"))
             |> Filter.build() == "artist==Radiohead;duration=gt=200000,genre==house"
    end
  end

  describe "build/1" do
    test "accepts a plain binary as identity" do
      assert Filter.build("artist==Radiohead") == "artist==Radiohead"
    end
  end
end

File: sdk/typescript/src/filter.test.ts
import { describe, expect, test } from "bun:test";
import { AlbumFields, ArtistFields, Filter, ScrobbleFields, SongFields } from "./filter.js";

describe("Filter", () => {
  test("eq renders a bare value when safe", () => {
    expect(Filter.eq("artist", "Radiohead").build()).toBe("artist==Radiohead");
  });

  test("eq quotes values with spaces", () => {
    expect(Filter.eq("artist", "Daft Punk").build()).toBe('artist=="Daft Punk"');
  });

  test("eq escapes embedded quotes and backslashes", () => {
    expect(Filter.eq("title", 'He said "hi"').build()).toBe('title=="He said \\"hi\\""');
    expect(Filter.eq("title", "back\\slash").build()).toBe('title=="back\\\\slash"');
  });

  test("wildcards pass through unquoted", () => {
    expect(Filter.eq("artist", "Daft*").build()).toBe("artist==Daft*");
  });

  test("ne", () => {
    expect(Filter.ne("artist", "Eminem").build()).toBe("artist!=Eminem");
  });

  test("ordered comparisons render verbose operators", () => {
    expect(Filter.gt("duration", 200000).build()).toBe("duration=gt=200000");
    expect(Filter.ge("year", 2000).build()).toBe("year=ge=2000");
    expect(Filter.lt("trackNumber", 5).build()).toBe("trackNumber=lt=5");
    expect(Filter.le("year", 1999).build()).toBe("year=le=1999");
  });

  test("in and out lists", () => {
    expect(Filter.in("genre", ["house", "electro"]).build()).toBe("genre=in=(house,electro)");
    expect(Filter.out("genre", ["hip hop"]).build()).toBe('genre=out=("hip hop")');
  });

  test("in/out reject empty lists", () => {
    expect(() => Filter.in("genre", [])).toThrow();
    expect(() => Filter.out("genre", [])).toThrow();
  });

  test("null checks", () => {
    expect(Filter.isNull("uri").build()).toBe("uri==null");
    expect(Filter.isNotNull("uri").build()).toBe("uri!=null");
  });

  test("and joins with ;", () => {
    const f = Filter.eq("artist", "Radiohead").and(Filter.gt("duration", 200000));
    expect(f.build()).toBe("artist==Radiohead;duration=gt=200000");
  });

  test("or joins with ,", () => {
    const f = Filter.eq("artist", "Radiohead").or(Filter.eq("artist", "Muse"));
    expect(f.build()).toBe("artist==Radiohead,artist==Muse");
  });

  test("or nested in and is parenthesized", () => {
    const left = Filter.eq("artist", "Radiohead").or(Filter.eq("artist", "Muse"));
    expect(left.and(Filter.gt("duration", 200000)).build()).toBe(
      "(artist==Radiohead,artist==Muse);duration=gt=200000",
    );
    const right = Filter.eq("genre", "house").or(Filter.eq("genre", "electro"));
    expect(Filter.eq("artist", "Radiohead").and(right).build()).toBe(
      "artist==Radiohead;(genre==house,genre==electro)",
    );
  });

  test("and nested in or needs no parentheses", () => {
    const f = Filter.eq("artist", "Radiohead")
      .and(Filter.gt("duration", 200000))
      .or(Filter.eq("genre", "house"));
    expect(f.build()).toBe("artist==Radiohead;duration=gt=200000,genre==house");
  });

  test("booleans and dotted fields", () => {
    expect(Filter.eq("user.handle", "tsiry.dev").build()).toBe("user.handle==tsiry.dev");
    expect(Filter.eq("liked", true).build()).toBe("liked==true");
  });

  test("toString matches build so filters interpolate", () => {
    const f = Filter.eq("artist", "Radiohead");
    expect(`${f}`).toBe(f.build());
  });

  test("field constants map to server selectors", () => {
    expect(SongFields.albumArtist).toBe("albumArtist");
    expect(AlbumFields.releaseDate).toBe("releaseDate");
    expect(ArtistFields.genres).toBe("genres");
    expect(ScrobbleFields.trackArtist).toBe("track.artist");
    expect(ScrobbleFields.userHandle).toBe("user.handle");
    expect(Filter.eq(ScrobbleFields.artistGenres, "house").build()).toBe("artist.genres==house");
  });
});
